src/analysis/p_series.lean

```diff
@@ -0,0 +1,227 @@
+  (∑' k, f k) ≤ f 0 + ∑' k : ℕ, (2 ^ k) * f (2 ^ k) :=
+begin
+  rw [ennreal.tsum_eq_supr_nat' (nat.tendsto_pow_at_top_at_top_of_one_lt _root_.one_lt_two)],
+  refine supr_le (λ n, (finset.le_sum_condensed hf n).trans (add_le_add_left _ _)),
+  simp only [nsmul_eq_mul, nat.cast_pow, nat.cast_two],
+  apply ennreal.sum_le_tsum
+end
+
+lemma tsum_condensed_le (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) :
+  (∑' k : ℕ, (2 ^ k) * f (2 ^ k)) ≤ f 1 + 2 * ∑' k, f k :=
+begin
+  rw [ennreal.tsum_eq_supr_nat' (tendsto_at_top_mono nat.le_succ tendsto_id), two_mul, ← two_nsmul],
+  refine supr_le (λ n, le_trans _ (add_le_add_left (nsmul_le_nsmul_of_le_right
+    (ennreal.sum_le_tsum $ finset.Ico 2 (2^n + 1)) _) _)),
+  simpa using finset.sum_condensed_le hf n
+end
+
+end ennreal
+
+namespace nnreal
+
+/-- Cauchy condensation test for a series of `nnreal` version. -/
+lemma summable_condensed_iff {f : ℕ → ℝ≥0} (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) :
+  summable (λ k : ℕ, (2 ^ k) * f (2 ^ k)) ↔ summable f :=
+begin
+  simp only [← ennreal.tsum_coe_ne_top_iff_summable, ne.def, not_iff_not, ennreal.coe_mul,
+    ennreal.coe_pow, ennreal.coe_two],
+  split; intro h,
+  { replace hf : ∀ m n, 1 < m → m ≤ n → (f n : ennreal) ≤ f m :=
+      λ m n hm hmn, ennreal.coe_le_coe.2 (hf (zero_lt_one.trans hm) hmn),
+    simpa [h, ennreal.add_eq_top] using (ennreal.tsum_condensed_le hf) },
+  { replace hf : ∀ m n, 0 < m → m ≤ n → (f n : ennreal) ≤ f m :=
+      λ m n hm hmn, ennreal.coe_le_coe.2 (hf hm hmn),
+    simpa [h, ennreal.add_eq_top] using (ennreal.le_tsum_condensed hf) }
+end
+
+end nnreal
+
+/-- Cauchy condensation test for series of nonnegative real numbers. -/
+lemma summable_condensed_iff_of_nonneg {f : ℕ → ℝ} (h_nonneg : ∀ n, 0 ≤ f n)
+  (h_mono : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) :
+  summable (λ k : ℕ, (2 ^ k) * f (2 ^ k)) ↔ summable f :=
+begin
+  set g : ℕ → ℝ≥0 := λ n, ⟨f n, h_nonneg n⟩,
+  have : f = λ n, g n := rfl,
+  simp only [this],
+  have : ∀ ⦃m n⦄, 0 < m → m ≤ n → g n ≤ g m := λ m n hm h, nnreal.coe_le_coe.2 (h_mono hm h),
+  exact_mod_cast nnreal.summable_condensed_iff this
+end
+
+open real
+
+/-!
+### Convergence of the `p`-series
+
+In this section we prove that for a real number `p`, the series `∑' n : ℕ, 1 / (n ^ p)` converges if
+and only if `1 < p`. There are many different proofs of this fact. The proof in this file uses the
+Cauchy condensation test we formalized above. This test implies that `∑ n, 1 / (n ^ p)` converges if
+and only if `∑ n, 2 ^ n / ((2 ^ n) ^ p)` converges, and the latter series is a geometric series with
+common ratio `2 ^ {1 - p}`. -/
+
+/-- Test for congergence of the `p`-series: the real-valued series `∑' n : ℕ, (n ^ p)⁻¹` converges
+if and only if `1 < p`. -/
+@[simp] lemma real.summable_nat_rpow_inv {p : ℝ} : summable (λ n, (n ^ p)⁻¹ : ℕ → ℝ) ↔ 1 < p :=
+begin
+  cases le_or_lt 0 p with hp hp,
+  /-- Caucy condensation test applies only to monotonically decreasing sequences, so we consider the
```

Comment:
I'm a little surprised that Lean accepts doc string comments in the middle of a proof:
```suggestion
  /- Cauchy condensation test applies only to monotonically decreasing sequences, so we consider the
```